### PR TITLE
Extend interfaces & performance optimization for KNN/DF/SVM algorithms. From 2020-u3-rls branch

### PR DESCRIPTION
**Pull request by OnlyDeniko**
Taken PR from master: #762(master), #910, #765, #857, #776, #924, #916, #921, #914, #942, #940, #941, #944, #955

### CODE REVIEW

**Comment by PetrovKP** [comment]
@agorshk check please changes for DF. @Alexsandruss  check please changes for KNN

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

File: .ci/scripts/conformance-scripts/download_tests.sh [hunk 1]
```diff
@@ -21,31 +21,37 @@ SKLEARN_PATH="$CONDA_PREFIX/lib/python${PYTHON_VERSION}/site-packages/sklearn/"
...
     ;;
     "svm_sparse")
-        cp ${SKLEARN_PATH}svm/tests/test_sparse.py test_svm_sparse.py
+        wget -O test_svm_sparse.py ${SKLEARN_URL_ROOT}svm/tests/test_sparse.py
+    ;;
+    "forest")
+        wget -O test_forest.py ${SKLEARN_URL_ROOT}ensemble/tests/test_forest.py
+    ;;
+    "knn")
+        wget -O test_knn.py ${SKLEARN_URL_ROOT}neighbors/tests/test_neighbors.py
```
Comment (Pahandrovich) [inline .ci/scripts/conformance-scripts/download_tests.sh:54]: In this file (download_tests.sh), need only to add copy test files for forest and knn.
```suggestion
    "forest")
        cp ${SKLEARN_PATH}ensemble/tests/test_forest.py test_forest.py
    ;;
    "knn")
        cp ${SKLEARN_PATH}neighbors/tests/test_neighbors.py test_knn.py
```

File: .ci/scripts/conformance_test.sh [hunk 2]
```diff
@@ -52,5 +52,5 @@ export LD_LIBRARY_PATH=${BUILD_DIR}/tbb/latest/lib/intel64:$LD_LIBRARY_PATH
 # testing
 cd .ci/scripts/conformance-scripts/ || exit 1
 export IDP_SKLEARN_VERBOSE=INFO
-python run_tests.py ${PYTHON_VERSION}
+python run_tests.py 0.23.2
```
[the remark this answers is not in the document]
Comment (Pahandrovich) [reply .ci/scripts/conformance_test.sh:55]:
```suggestion
python run_tests.py ${PYTHON_VERSION}
```

File: .ci/scripts/conformance-scripts/download_tests.sh (hunk 1 above)
Comment (Pahandrovich) [reply .ci/scripts/conformance-scripts/download_tests.sh:54]: For other algorithms leave it as it was before the changes

Comment (OnlyDeniko) [reply .ci/scripts/conformance-scripts/download_tests.sh]: Okay, corrected

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by agorshk** [comment]
The file with name df_training_parameter.cpp doesn't in the list of file changes. Please have a look [PR](https://github.com/oneapi-src/oneDAL/pull/681) In file df_training_parameter.cpp: seed(777)

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/test_branch

**Review by agorshk — COMMENTED** [review]
Ok, from side of Decision Forest. Approve for other algorithms are needed

**Review by PetrovKP — APPROVED** [review]
From my side - Ok

**Comment by OnlyDeniko** [comment]
/intelci: run
set infra_branch=dev/dkulandi/2021-mnt

**Comment by PetrovKP** [comment]
/intelci: restart
set infra_branch=dev/dkulandi/2021-mnt

**Comment by Alexsandruss** [comment]
Cherry pick https://github.com/oneapi-src/oneDAL/pull/991 after it's merge.

**Comment by OnlyDeniko** [comment]
/intelci: restart
set infra_branch=dev/dkulandi/2021-mnt

**Comment by OnlyDeniko** [comment]
/intelci: restart
set infra_branch=dev/dkulandi/2021-mnt

**Review by Alexsandruss — APPROVED** [review]

**Pull request merged**